tests/SchedulerTest.php

```diff
@@ -0,0 +1,221 @@
+<?php
+
+namespace ipl\Tests\Scheduler;
+
+use DateTime;
+use ipl\Scheduler\Contract\Task;
+use ipl\Tests\Scheduler\Lib\AbsoluteDueFrequency;
+use ipl\Tests\Scheduler\Lib\CountableScheduler;
+use ipl\Tests\Scheduler\Lib\ImmediateDueFrequency;
+use ipl\Tests\Scheduler\Lib\NeverDueFrequency;
+use ipl\Tests\Scheduler\Lib\PromiseBoundTask;
+use ipl\Tests\Scheduler\Lib\TaskRejectedException;
+use PHPUnit\Framework\TestCase;
+use React\EventLoop\Loop;
+use React\Promise;
+use React\Promise\ExtendedPromiseInterface;
+use Throwable;
+
+class SchedulerTest extends TestCase
+{
+    /** @var CountableScheduler */
+    protected $scheduler;
+
+    public function setUp(): void
+    {
+        $this->scheduler = new CountableScheduler();
+    }
+
+    protected function runOff()
+    {
+        Loop::futureTick(function () {
+            Loop::stop();
+        });
+
+        Loop::run();
+    }
+
+    public function testSchedulingTasks()
+    {
+        $task = new PromiseBoundTask(Promise\resolve());
+        $now = new DateTime();
+        $now->modify('+2 month');
+
+        $nextDue = null;
+        $this->scheduler
+            ->on(CountableScheduler::ON_TASK_SCHEDULED, function (Task $_, DateTime $time) use (&$nextDue) {
+                $nextDue = $time;
+            })
+            ->schedule($task, new AbsoluteDueFrequency($now));
+
+        $this->runOff();
+
+        $this->assertEquals($now, $nextDue, 'Scheduler could not get the correct schedule date time');
+    }
+
+    public function testNeverDueTaskSchedules()
+    {
+        $neverRun = true;
+        $this->scheduler
+            ->schedule(new PromiseBoundTask(Promise\resolve()), new NeverDueFrequency())
+            ->on(CountableScheduler::ON_TASK_RUN, function (Task $_, ExtendedPromiseInterface $p) use (&$neverRun) {
+                $neverRun = false;
+            });
+
+        $this->runOff();
+
+        $this->assertTrue($neverRun, 'Scheduler::schedule() runs a task with a never due frequency');
+    }
+
+    public function testRunningScheduledTasks()
+    {
+        $hasRun = false;
+        $task = new PromiseBoundTask(Promise\resolve());
+        $this->scheduler
+            ->schedule($task, new ImmediateDueFrequency())
+            ->on(CountableScheduler::ON_TASK_RUN, function (Task $t, ExtendedPromiseInterface $_) use (&$hasRun) {
+                $hasRun = true;
+            });
+
+        $this->runOff();
+
+        $this->assertTrue($hasRun, 'Scheduler::schedule() could not run scheduled tasks');
+    }
+
+    public function testCancelingRunningTasks()
```

Comment:
Can you also add a test for canceling scheduled tasks that are not currently running?